### PR TITLE
Implemented loading custom GPT2LMHeadModel from _model_path

### PR DESCRIPTION
**Pull request by geranium12**
## Summary
While testing my modified GPT2LMHeadModel, I encountered some issues loading the model to the app as a HookedTransformer. Subsequently, I integrated the necessary changes to be able to load this custom model supported by HuggingFace, but, for example, has a different number of layers, heads, etc. compared to the pretrained version. This also allows one to easily integrate other models into the app.

## Implemented Changes
- Added llm_transparency_tool/models/load_custom_model.py
    - It includes a function converting a HuggingFace model, e.g. GPT2LMHeadModel, to a HookedTransformer from transformer_lens.
- Modified llm_transparency_tool/models/tlens_model.py
   - Added an entry to hash_func enabling Streamlit to cache the hf_model variable.
   - Changed max_entries from 1 to 2 to store tlens_model and hf_model in cache together.
    - Now if hf_model is loaded, tlens_model is loaded via the hf_model_to_hooked_transformer function.
- Modified llm_transparency_tool/server/utils.py
    - Made use of _model_path using it for loading GPT2LMHeadModel and its tokenizer. Nevertheless, it doesn't check if the stored model in _model_path is of that class.

### CODE REVIEW

**Review by esetono1994 — APPROVED** [review]

File: llm_transparency_tool/server/utils.py
Comment (tsaoyu) [inline llm_transparency_tool/server/utils.py:65]: But this change breaks other thing if model is not `GPT2LMHeadModel`

Comment (geranium12) [reply llm_transparency_tool/server/utils.py]: The problem is that TransformerLens only supports the loading of pre-trained models. If a huggingface model has, for instance, a different number of layers, TransformerLens is unable to load it. In such cases, a user has to implement the loading of the model to HookedTransformer in hf_model_to_hooked_transformer() on their own. Since we don't know if the pre-trained or custom model needs to be loaded from _model_path, I'd rather leave it to the user to modify the code if necessary. Adding support for other models is now easier by looking at how I implemented it for GPT2LMHeadModel. Anyway, I've made a few changes to be able to read any model from _model_path.

**Comment by mahnerak** [comment]
Thank you, @geranium12, for your interest and for contributing.

I'll need some time to review and ensure compatibility with the upcoming changes from my side (that also include changes to support local paths) before merging.

File: llm_transparency_tool/models/hash_funcs.py
Comment (tsaoyu) [inline llm_transparency_tool/models/hash_funcs.py:9]: I also found this limitation annoying, this means we have to run into hash function error then add them back to here.  Which means this list will getting longer and longer when supporting more models, even for models loaded from local directory.

File: llm_transparency_tool/models/tlens_model.py
Comment (tsaoyu) [inline llm_transparency_tool/models/tlens_model.py:47]: I did a monkey patch to the upstream `TransformerLens` for this. For the separation of concern, support more model arch is the responsibility of that library. `LLM-TT` should deal with right configuration to load from local directory and accepting custom model arch if that is accepted by `TransformerLens`.

**Comment by mahnerak** [comment]
Thanks @geranium12 and @tsaoyu. I agree with both of your points.
I've just pushed changes to our approach, which is preferred because it makes it easier to use the UI app / tool with local models **without code changes**. Please find the instructions in #28.

Briefly, the `supported_model_name` parameter has been added, which selects the model conversion and hooks logic to be used by TransformerLens. This eliminates the need to hardcode the model classes.
During loading, two arguments can be passed:
- The model's local path, identifier, or URL, passed with `model_name`
- The compatible model name, passed with `supported_model_name`
I believe this is easier, as the user is very likely to know the base model that was fine-tuned.

**Pull request closed without merging**
